tests/unit/test_tiler.py

```diff
@@ -1253,3 +1253,376 @@ def it_can_extract_score_tiles_and_save_report(self, request, tmpdir):
             [(0.8, coords), (0.7, coords)],
             [f"tile_{i}_level2_0-0-10-10.png" for i in range(2)],
         )
+
+
+class Describe_ThresholdTiler:
+    def it_constructs_from_args(self, request):
+        _init = initializer_mock(request, ThresholdTiler)
+        rs = RandomScorer()
+        threshold_tiler = ThresholdTiler(
+            rs, (512, 512), 0.1, 2, True, 80, 0, "", ".png", mpp=None
+        )
+
+        _init.assert_called_once_with(
+            ANY, rs, (512, 512), 0.1, 2, True, 80, 0, "", ".png", mpp=None
+        )
+
+        assert isinstance(threshold_tiler, ThresholdTiler)
+        assert isinstance(threshold_tiler, GridTiler)
+        assert isinstance(threshold_tiler, Tiler)
+
+    def it_knows_its_scorer(self):
+        random_scorer = RandomScorer()
+        threshold_tiler = ThresholdTiler(random_scorer, (512, 512), 0.1, 0)
+
+        scorer_ = threshold_tiler.scorer
+
+        assert callable(scorer_)
+        assert isinstance(scorer_, RandomScorer)
+
+    def it_knows_its_threshold(self):
+        threshold = 0.1
+        threshold_tiler = ThresholdTiler(RandomScorer(), (512, 512), threshold, 0)
+
+        threshold_ = threshold_tiler.threshold
+
+        assert type(threshold_) == float
+        assert threshold_ == threshold
+
+    @pytest.mark.parametrize(
+        "tile_size, expected_result", [((512, 512), False), ((200, 200), True)]
+    )
+    def it_knows_if_it_has_valid_tile_size(self, tmpdir, tile_size, expected_result):
+        slide, _ = base_test_slide(tmpdir, PILIMG.RGBA_COLOR_500X500_155_249_240)
+        threshold_tiler = ThresholdTiler(RandomScorer(), tile_size, 0.2, 0)
+
+        result = threshold_tiler._has_valid_tile_size(slide)
+
+        assert type(result) == bool
+        assert result == expected_result
+
+    def it_can_calculate_scores(self, request):
+        slide = instance_mock(request, Slide)
+        coords = CP(0, 0, 10, 10)
+        image = PILIMG.RGB_RANDOM_COLOR_500X500
+        tile = Tile(image, coords)
+        _tiles_generator = method_mock(request, GridTiler, "_tiles_generator")
+        # it needs to be a generator
+        _tiles_generator.return_value = ((tile, coords) for i in range(3))
```

Comment:
The `_tiles_generator` return value in the signatures is `Tuple[List[Tuple[float, CoordinatePair]], List[Tuple[float, CoordinatePair]]]` and here you are assigning `(tile, coords)` I don't think it's quite correct.
So or the test is wrong or the signature is wrong. :D

Reply:
The method being mocked is `_tiles_generator` of the `GridTiler` class, which I think has the signature `Tuple[Tile, CoordinatePair]`, so I think both are correct (?)

Reply:
> Oh ok, but now the question is why are you mocking the `GridTiler`?

Because the aim of the test is to check whether the tiler can extract the scores from tiles, so I think it does not matter if the tiles come from a `GridTiler`.

In the test `it_can_calculate_filtered_tiles` I am calling the `_tiles_generator` from a `ThresholdTiler`.

Reply:
I agree with @ernestoarbitrio: here we should mock the `_tiles_generator` from the `ThresholdTiler` to be more precise (and the tests for the `ScoreTiler` should do the same, oops) and assign the correct return value